### ISSUE TITLE
[DataGridPro] Allow to use `valueGetter` on the tree data grouping column

### ISSUE THREAD

**Issue by TrevorThomp** [mui/mui-x#4035]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

It would be nice to have the ability to have the `getTreeDataPath` to be targeting unique values such as IDs. However, I don't want them rendered into the grouping column rows and would like them to be a different value (e.g. names). Currently using names has the possibility of throwing an error because `getTreeDataPath` is expecting unique values.

In the docs, if you were to edit one of the child names to equal another you get the following error:

```MUI: The path returned by `getTreeDataPath` should be unique.```

https://mui.com/components/data-grid/group-pivot/#group-expansion-2
https://codesandbox.io/s/treedatawithgap-material-demo-forked-ly3lct?file=/demo.js

The ultimate goal would be the same tree data layout like the names are used in the docs, but using a value that is targeted to be more unique.

PR referencing something similar:
https://github.com/mui/mui-x/issues/3672

### Examples 🌈

ValueGetter
```jsx
const rows = [
  {
    id: 1,
    hierarchy: [1],
   'Robert'
  },
  {
   id: 2,
   hierarchy: [1, 2]
   'Karen'
  },
  {
   id: 3,
   hierarchy: [1, 3]
   name: 'Ted'
  },
];

<DataGridPro
  rows={rows}
  treeData={true}
  getTreeDataPath={(row) => row.hierarchy}
  groupingColDef={{ 
   headerName: "Name" 
   valueGetter: ({ row }) => {
     return row.name;
   }
  }}
/>
```

### Motivation 🔦

Ensuring that the tree data is basing its pathing off of unique values and being able to render different values in the cells like we are able to do in other columns.

**Comment by flaviendelangle** [comment]
Hi, thanks for your suggestion.
I have a few questions to make sure that we are on the same page.

In your scenario, would the sorting and filtering be done based on the "groupingKey" (ie the value of the returned by `getTreeDataPath` that is used to group the children of the current row) or on the non-uniq value you are describing ?
Depending on that, what you need it either `valueGetter` or `valueGetter` + `valueFormatter`.

`valueFormatter` has been implemented in #4022 and will be released this week or next week I think.
[Here is a working example](https://codesandbox.io/s/treedatawithgap-material-demo-forked-vyvnlx?file=/demo.js). Not that the filtering / sorting is still done based on your IDs.

`valueGetter` works properly for the Row Grouping feature, but has not been implemented for the Tree Data. If that's something you need, I can plan to add it.

**Comment by flaviendelangle** [comment]
I thought it would not work, but it seems that can just pass the `valueGetter` like [here](https://codesandbox.io/s/treedatawithgap-material-demo-forked-kcmj58?file=/demo.js)

**Comment by TrevorThomp** [comment]
@flaviendelangle I may be misunderstanding you. Is that supposed to be working in the latest release or you thought the functionality wouldn't work? The demo looks great!

**Comment by flaviendelangle** [comment]
I thought I needed to do some changes
But it seems to work on the currently published version

**Comment by TrevorThomp** [comment]
@flaviendelangle I'm not sure if that is the case. I'm unable to get it to work locally and the `package.json` for that demo was pointing to a specific commit. I updated it to point to the latest published version and it no longer displays the name as expected [Demo Here](https://codesandbox.io/s/treedatawithgap-material-demo-forked-41q1hd?file=/demo.js)

When looking at the id's it looks like it's sorting them off the `valueGetter` correctly, but not displaying them inside the cells and still using the `groupingKey`

**Comment by flaviendelangle** [comment]
The `valueFormatter` improvement has been done in #4022, which will be released in 5.6 :+1:

**Comment by TrevorThomp** [comment]
@flaviendelangle Perfect! I will wait for the release. Thanks for your help on this
